src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -4508,5 +4510,206 @@ public void GetRelatedDocumentsDoesNotReturnOtherTypesOfDocuments()
+            Assert.Single(frozenCompilation2.SyntaxTrees);
+            Assert.True(frozenCompilation2.ContainsSyntaxTree(await frozenProject2.Documents.Single().GetSyntaxTreeAsync()));
+
+            Assert.Single(frozenCompilation2.References.Where(r => r is CompilationReference c && c.Compilation == frozenCompilation1));
+        }
+
+        [Fact]
+        public async Task TestFrozenPartialSolution6()
+        {
+            using var workspace = WorkspaceTestUtilities.CreateWorkspaceWithPartialSemantics();
+            var project1 = workspace.CurrentSolution.AddProject("CSharpProject1", "CSharpProject1", LanguageNames.CSharp);
+            project1 = project1.AddDocument("Doc1", SourceText.From("class Doc1 { }")).Project;
+
+            // If we freeze
+            var compilation1 = await project1.GetCompilationAsync();
+            var syntaxTree1 = await project1.Documents.Single().GetSyntaxTreeAsync();
+
+            var frozenSolution = await project1.Solution.WithFrozenPartialCompilationsAsync(CancellationToken.None);
+
+            var forkedProject1 = frozenSolution.WithDocumentText(project1.Documents.Single().Id, SourceText.From("class Doc2 { }")).GetProject(project1.Id);
+            var forkedDocument1 = forkedProject1.Documents.Single();
+            var forkedSyntaxTree1 = await forkedDocument1.GetSyntaxTreeAsync();
+
+            Assert.NotEqual(syntaxTree1, forkedSyntaxTree1);
+
+            var forkedCompilation1 = await forkedProject1.GetCompilationAsync();
+            Assert.NotEqual(compilation1, forkedCompilation1);
+            Assert.True(forkedCompilation1.ContainsSyntaxTree(forkedSyntaxTree1));
+        }
+
+        [Theory, CombinatorialData]
+        public async Task TestFrozenPartialSolution7(bool freeze)
+        {
+            using var workspace = WorkspaceTestUtilities.CreateWorkspaceWithPartialSemantics();
+            var project1 = workspace.CurrentSolution.AddProject("CSharpProject1", "CSharpProject1", LanguageNames.CSharp);
+            project1 = project1.AddDocument("Doc1", SourceText.From("class Doc1 { }")).Project;
+
+            var invokeIndex = 1;
+            project1 = project1.AddAnalyzerReference(new TestGeneratorReference(new CallbackGenerator(() =>
+            {
+                var index = invokeIndex++;
+                return ("hintname" + index, "// source" + index);
+            })));
+
+            var compilation1 = await project1.GetCompilationAsync();
+            var syntaxTree1 = await project1.Documents.Single().GetSyntaxTreeAsync();
+            var generatedDocuments = await project1.GetSourceGeneratedDocumentsAsync();
+
+            Assert.Single(generatedDocuments);
+            Assert.Equal("// source1", generatedDocuments.Single().GetTextSynchronously(CancellationToken.None).ToString());
+
+            var frozenSolution = freeze
+                ? await project1.Solution.WithFrozenPartialCompilationsAsync(CancellationToken.None)
+                : project1.Solution;
```

Comment:
Why is the testing in the non-frozen case happening here? Is that just a generic test of generators-functioning-like-they should to avoid the interesting case passing in a way we didn't expect?

Reply:
i liked just validating what happens there.  It also shows we absolutely expect a difference in behavhior, and that we're not just inheriting the non-frozen behavior which works differently than you or i might be expecting.